lib/CsvImportValidator.class.php

```diff
@@ -0,0 +1,320 @@
+    {
+      if (false === $data = fread($fh, strlen($value)))
+      {
+        throw new sfException('Failed to read from CSV file in handleByteOrderMark.');
+      }
+
+      if (0 === strncmp($data, $value, strlen($value)))
+      {
+        return; // Just eat the BOM and move on from this file position
+      }
+
+      // No BOM, rewind the file handle position
+      if (false === rewind($fh))
+      {
+        throw new sfException('Rewinding file position failed in handleByteOrderMark.');
+      }
+    }
+  }
+
+  public function loadCsvData($fh)
+  {
+    $this->handleByteOrderMark($fh);
+
+    $this->columnNames = fgetcsv($fh, 60000);
+
+    if ($this->columnNames === false)
+    {
+      throw new sfException('Could not read initial row. File could be empty.');
+    }
+
+    while ($item = fgetcsv($fh, 60000))
+    {
+      $this->rows[] = $item;
+    }
+  }
+
+  public function validate()
+  {
+    if (false === $fh = fopen($this->filename, 'rb'))
+    {
+      throw new sfException('You must specify a valid filename');
+    }
+
+    $this->loadCsvData($fh);
+
+    foreach ($this->csvTests as $test)
+    {
+      $test->setFilename($this->filename);
+    }
+
+    foreach ($this->rows as $row)
+    {
+      if ($this->showDisplayProgress)
+      {
+        print $this->renderProgressDescription();
+      }
+      
+      foreach ($this->csvTests as $test)
+      {
+        $test->testRow(array_combine($this->columnNames, $row));
+      }
+    }
+
+    if ($this->showDisplayProgress)
+    {
+      print $this->renderProgressDescription(true);
+    }
+  }
+
+  public function setShowDisplayProgress(bool $value)
+  {
+    $this->showDisplayProgress = $value;
+  }
+
+  public function getContext()
+  {
+    return $this->context;
+  }
+
+  public function setContext($context)
+  {
+    $this->context = $context;
+  }  
+
+  // Collect all the results from all CSV Validations.
+  // Return array of results sorted by filename.
+  public function getResults()
+  {
+    $results = array();
+
+    foreach ($this->getCsvTests() as $test)
+    {
+      $results[$test->getFileName()][] = $test->getTestResult();
+    }
+    
+    //var_dump($results);
+    return $results;
+  }
+
+  public function setCsvTests(array $classes)
+  {
+    unset($this->csvTests);
+
+    foreach($classes as $key => $class)
+    {
+      $this->csvTests[] = new $class();  
+    }
+  }
+
+  public function getCsvTests()
+  {
+    return $this->csvTests;
+  }
+
+  public function setOptions(array $options = null)
+  {
+    if (empty($options))
+    {
+      return;
+    }
+
+    foreach ($options as $name => $val)
+    {
+      $this->setOption($name, $val);
+    }
+  }
+
+  public function setImportType(string $value)
+  {
+    if (in_array($value, $this->defaultCsvTypeMap))
+    {
+      $this->validatorOptions['importType'] = $value;
+    }
+    else
+    {
+      throw new UnexpectedValueException(sprintf('Invalid option "%s".', $name));
```

Comment:
Should it be "Invalid name"? Where is $name set?

Reply:
Good catch! Fixed...